Support Debug Console interacting directly with GDB

trongle0504 opened the pull request:
This change adds the ability to interact with GDB directly in the debug console of VSCode by providing a way to bypass the normal expression evaluation and sending commands to the GDB console input.

To display the value of an expression, type that expression which can reference variables that are in scope. For example type `2 + 3` or the name of a variable. Arbitrary commands can be sent to GDB by prefixing the input with a `>`, for example type `>show version` or `>help`.

trongle0504 commented:
Hi @jonahgraham,
I have added some code to support sending GDB commands on CLI as your suggestion before.
Could you please help me review it?
Thank you!

jonahgraham reviewed (commented):
Overall this LGTM. Please review test failures  Note until your first PR is approved I need to approve each run on the test suite in the PR, so please check the actions are successful in your own fork (https://github.com/trongle0504/cdt-gdb-adapter/actions)

trongle0504 commented on src/GDBDebugSession.ts:
```diff
@@ -1232,6 +1232,20 @@ export class GDBDebugSession extends LoggingDebugSession {
             }
+
+            if (args.expression.startsWith('>')) {
+                await mi.sendThreadSelectRequest(this.gdb, {
```
I updated it

trongle0504 commented on src/integration-tests/evaluate.spec.ts:
```diff
@@ -79,4 +79,41 @@ describe('evaluate request', function () {
 
         expect(err.message).eq('-var-create: unable to create variable object');
     });
+    it('should be able to update value of a variable which has local scope and named "monitor"', async function () {
```
I updated it

trongle0504 commented:
Hi @jonahgraham,
I have some updates as your suggestion.  The unit test also passed in my fork.
Could you please help me review it again?
Thank you!

jonahgraham commented on src/GDBDebugSession.ts:
```diff
@@ -1232,6 +1232,21 @@ export class GDBDebugSession extends LoggingDebugSession {
             }
+
+            if (args.expression.startsWith('>') && args.context === 'repl') {
+                await mi.sendInterpreterExecConsole(this.gdb, {
```
I think you have misinterpreted the use of interpreter-exec console. Please see https://sourceware.org/gdb/onlinedocs/gdb/GDB_002fMI-Miscellaneous-Commands.html#index-_002dinterpreter_002dexec and https://sourceware.org/gdb/onlinedocs/gdb/Context-management.html#Context-management

jonahgraham replied:
PS I don't know whether it is better to use interpreter-exec or not, I know that is how Eclipse CDT does it to though.

trongle0504 replied:
Hi @jonahgraham,
Thanks a lot for your documentation. I have some updates in the codes. I compared the currents response at CLI in VS code and e2studio that they are similar. Could you please take a look the changes? If there is anything that need to be updated, please let me know.

jonahgraham reviewed (changes requested):
This is very close to right, please review my latest comments.

trongle0504 commented on src/integration-tests/evaluate.spec.ts:
```diff
@@ -79,4 +79,46 @@ describe('evaluate request', function () {
+        expect(res1.body.result).eq('\r');
+        const res2 = await dc.evaluateRequest({
+            context: 'repl',
+            expression: '>-gdb-version',
```
> `repl` is for humans to use, so no need to support MI.

I just based on e2studio to implement in VS code. On e2studio, user also can use MI command while debugging . If we don't need to support for MI command, I will update it.

trongle0504 replied:
> Why enable this flow? For starters `-` isn't [the only way](https://sourceware.org/gdb/onlinedocs/gdb/GDB_002fMI-Input-Syntax.html#GDB_002fMI-Input-Syntax) to type an MI command, e.g. `123-gdb-version` is also an MI command.

I just based on what e2studio implemented. You can take a look at below image.  For commands have "-" the interpreter will implicitly understand that is a MI command. Others cases will not be a MI command then it will add more `-interpreter-exec console` before send to GDB
![image](https://user-images.githubusercontent.com/104537989/235020455-b319e3a2-63d8-4cc9-9a71-936acc37f3b7.png)

trongle0504 replied:
Currently, in e2studio the interpreter relies on the first character to determine whether the command from the user is cli-command or mi-command. If the first character is a flash (-) the interpreter will understand implicitly that is a MI command, otherwise it's cli-command. 
![image](https://user-images.githubusercontent.com/104537989/235026378-6f93ae59-a93d-451d-aaf1-92b1177fa413.png)

trongle0504 replied:
> For starters `-` isn't [the only way](https://sourceware.org/gdb/onlinedocs/gdb/GDB_002fMI-Input-Syntax.html#GDB_002fMI-Input-Syntax) to type an MI command, e.g. `123-gdb-version` is also an MI command.

`-gdb-version` also has the first character is a dash. So the interpreter will understand implicitly is a MI command.
![image](https://user-images.githubusercontent.com/104537989/235028930-ff89859d-a888-46bd-bc24-b36c308254cd.png)

trongle0504 replied:
Could you please help me confirm whether we should support MI command on cdt-gdb-adapter?

trongle0504 commented on src/mi/interpreter.ts:
```diff
@@ -0,0 +1,22 @@
+    }
+) {
+    return gdb.sendCommand(
+        `-interpreter-exec console "thread ${params.threadId}" "frame ${params.frameId}" "${params.command}"`
```
Sorry if I misunderstood your idea. Do you mean it should be `-interpreter-exec console "${params.command}" --frame ${params.frameId} --thread ${params.threadId} ` ?. If it is right, It's seem that `--frame` and `-- thread ` are not accepted the interpreter. I got error when send the command `-interpreter-exec console help --frame 0 --thread 1`, the error is `Undefined command: "--frame".  Try "help".` In e2studio, --frame and --thread info also don't  need include in GDB command when sending 
![image](https://user-images.githubusercontent.com/104537989/235016503-b8d0ca18-23bb-4b1b-ade1-130ae20f0bd9.png)

trongle0504 replied:
I have try to change the format command to `-interpreter-exec --thread ${params.threadId} --frame ${params.frameId} console "${params.command}` then it can work normally

jonahgraham commented on src/GDBDebugSession.ts:
```diff
@@ -1232,6 +1232,25 @@ export class GDBDebugSession extends LoggingDebugSession {
             }
+
+            if (args.expression.startsWith('>') && args.context === 'repl') {
+                if (args.expression[1] === '-') {
```
Please disregard the previous comment - too much reviewing java.

jonahgraham reviewed (approved):
This looks good to go now.

I have added an additional commit 0afeb61d318516691002ad66e5d0a0fec1dc570d to show the user some help when they startup. Can you please confirm this looks ok:

![image](https://user-images.githubusercontent.com/679236/235174329-ef93f265-2c89-4c5b-9b01-4728915aecc5.png)

jonahgraham commented:
> I have added an additional commit [0afeb61](https://github.com/eclipse-cdt-cloud/cdt-gdb-adapter/commit/0afeb61d318516691002ad66e5d0a0fec1dc570d) to show the user some help when they startup. Can you please confirm this looks ok:

I have split this out into #266

trongle0504 commented:
> > I have added an additional commit [0afeb61](https://github.com/eclipse-cdt-cloud/cdt-gdb-adapter/commit/0afeb61d318516691002ad66e5d0a0fec1dc570d) to show the user some help when they startup. Can you please confirm this looks ok:
> 
> I have split this out into #266

I got it. I think that your changes are very helpful.

jonahgraham commented:
Thanks @trongle0504 for the new feature.
